src/Foundatio/Jobs/IQueueJob.cs

```diff
@@ -18,6 +19,19 @@ public interface IQueueJob<T> : IJob where T : class
 
 public static class QueueJobExtensions
 {
+    public static async Task RunUntilEmptyAsync<T>(this IQueueJob<T> job, TimeSpan acquireTimeout,
+        CancellationToken cancellationToken = default) where T : class
+    {
+        if (acquireTimeout <= TimeSpan.Zero)
+            throw new ArgumentException("Acquire timeout must be greater than zero", nameof(acquireTimeout));
+
+        using var linkedCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+        linkedCancellationTokenSource.CancelAfter(acquireTimeout);
+
+        // NOTE: This has to be awaited otherwise the linkedCancellationTokenSource cancel timer will not fire.
+        await job.RunUntilEmptyAsync(linkedCancellationTokenSource.Token);
```

Comment:
NOTE: this was initially just returning the task, which did not dispose of the cancellation token but DID prevent the cancellation token timer from firing.... This was a huge time sink to track down.

Reply:
Might be something with https://github.com/microsoft/referencesource/blob/master/mscorlib/system/threading/CancellationTokenSource.cs#L460